### DISCUSSION TITLE
几种相对坐标转绝对坐标的实现方法

### DISCUSSION THREAD

**Discussion by I-love-study** [Xiaokang2022/maliang#35] [Ideas]
因为 `tkintertools` 是基于坐标绘制的 UI 库，所以需要很多绝对坐标
但是如何有些组件坐标关联程度较大的话，就可能会出现牵一发而动全身的效果
并且由于 `position` 的 type hint 要求是 `tuple[int, int]`，所以像是使用 `list`, `numpy.array` 之类的就会在类型检查的时候报红。
为了能够不要每次修改坐标都需要牵一发而动全身，下面有几种方法能够简化大家计算绝对路径的方法

## 1. 使用函数工厂

```python
def create_reletive_coords(x: int, y: int):

    def trans_coords(x_rele: int, y_rele: int):
        return (x + x_rele, y + y_rele)

    return trans_coords

rc = create_reletive_coords(450, 300)
tkt.Text(cv, rc(0, 0) , text="账号", anchor="nw")
tkt.InputBox(cv, rc(0, 40), (380, 50), placeholder="点击输入账号")
```

通过创建一个工厂函数，返回可以创建一个绝对坐标转相对坐标的函数

基于上面的理念，我们还可以创建一个类来实现更多功能

## 2. 类

```python
from typing import NamedTuple
import tkintertools as tkt


class ReletiveCoords(NamedTuple):
    x: int
    y: int

    def __add__(self, other: tuple[int, int]) -> "ReletiveCoords":
        return ReletiveCoords(self.x + other[0], self.y + other[1])

    __radd__ = __add__

    def __sub__(self, other: tuple[int, int]) -> "ReletiveCoords":
        return ReletiveCoords(self.x - other[0], self.y - other[1])

    def __rsub__(self, other: tuple[int, int]) -> "ReletiveCoords":
        return ReletiveCoords(other[0] - self.x, other[1] - self.y)

    def __call__(self, x: int, y: int) -> "ReletiveCoords":
        return ReletiveCoords(self.x + x, self.y + y)

rc = ReletiveCoords(450, 300)
tkt.Text(cv, rc , text="账号", anchor="nw")
tkt.InputBox(cv, rc(0, 40), (380, 50), placeholder="点击输入账号")

rc2 = rc + (0, 100)
tkt.Text(cv, rc2, text="密码", anchor="nw")
tkt.InputBox(cv, rc(0, 40), (380, 50), show="●", placeholder="点击输入密码")

```

通过让 NamedTuple 添加了 `__add__`, `__radd__`, `__sub__`, `__rsub__`, `__call__` 的魔法方法，以实现相关方法，相比于第一个方法，该方法**能够自由加减，以达到更好的使用便利性**

**Comment by Xiaokang2022** [comment]
@I-love-study

关于这篇讨论已被该项目收录至博客页了，如果您有其他需求（删除，修改等），请尽管提出来。若无其他要求，**请随意回复一些内容**，以便我知晓您已知晓，然后我将关闭该讨论，谢谢。

收录见：https://xiaokang2022.github.io/tkintertools/blog/2024/09/09/%E5%87%A0%E7%A7%8D%E7%9B%B8%E5%AF%B9%E5%9D%90%E6%A0%87%E8%BD%AC%E7%BB%9D%E5%AF%B9%E5%9D%90%E6%A0%87%E7%9A%84%E5%AE%9E%E7%8E%B0%E6%96%B9%E6%B3%95%E5%AE%98%E6%96%B9%E6%9A%82%E6%97%B6%E6%94%B6%E5%BD%95/

**Reply by I-love-study** [reply]
> @I-love-study
> 
> 关于这篇讨论已被该项目收录至博客页了，如果您有其他需求（删除，修改等），请尽管提出来。若无其他要求，**请随意回复一些内容**，以便我知晓您已知晓，然后我将删除该讨论，谢谢。
> 
> 收录见：https://xiaokang2022.github.io/tkintertools/blog/2024/09/09/%E5%87%A0%E7%A7%8D%E7%9B%B8%E5%AF%B9%E5%9D%90%E6%A0%87%E8%BD%AC%E7%BB%9D%E5%AF%B9%E5%9D%90%E6%A0%87%E7%9A%84%E5%AE%9E%E7%8E%B0%E6%96%B9%E6%B3%95%E5%AE%98%E6%96%B9%E6%9A%82%E6%97%B6%E6%94%B6%E5%BD%95/

该讨论还是有一些错别字在里面，目前看到的有
如何 -> 如果，如果可以，请帮我修改一下。
很荣幸能够被项目收录至博客页。

**Reply by Xiaokang2022** [reply]
好的，将于近期修改。

**Reply by Xiaokang2022** [reply]
@I-love-study 已修改，可见链接页查看，如果仍有不满意的地方，可以自行创建 PR 来修改

关于创建 PR 修改的说明见：https://xiaokang2022.github.io/tkintertools/blog/9999/09/09/%E7%BD%91%E7%AB%99%E6%96%B0%E5%A2%9E%E5%8D%9A%E5%AE%A2%E5%8A%9F%E8%83%BD/#%E4%B8%89%E5%A6%82%E4%BD%95%E6%8A%95%E7%A8%BF%E5%92%8C%E4%BF%AE%E6%94%B9
